### ISSUE TITLE
java.nio.BufferUnderflowException in BdbMultipleWorkQueues.get

### ISSUE THREAD

**Issue by anjackson** [internetarchive/heritrix3#278]
The issue that led to #227 was some errors that killed a few ToeThreads on restarting from a checkpoint:

```
SEVERE: org.archive.crawler.framework.ToeThread run Fatal exception in ToeThread #356 (in thread 'ToeThread #356') [Thu Aug 08 07:58:40 GMT 2019] 
com.esotericsoftware.kryo.SerializationException: Buffer limit exceeded reading object of type: org.archive.modules.CrawlURI
	at com.esotericsoftware.kryo.Kryo.readObjectData(Kryo.java:516)
	at com.esotericsoftware.kryo.ObjectBuffer.readObjectData(ObjectBuffer.java:212)
	at org.archive.bdb.KryoBinding.entryToObject(KryoBinding.java:84)
	at org.archive.crawler.frontier.BdbMultipleWorkQueues.get(BdbMultipleWorkQueues.java:267)
	at org.archive.crawler.frontier.BdbWorkQueue.peekItem(BdbWorkQueue.java:103)
	at org.archive.crawler.frontier.WorkQueue.peek(WorkQueue.java:174)
	at org.archive.crawler.frontier.WorkQueueFrontier.findEligibleURI(WorkQueueFrontier.java:633)
	at org.archive.crawler.frontier.AbstractFrontier.next(AbstractFrontier.java:457)
	at org.archive.crawler.framework.ToeThread.run(ToeThread.java:134)
Caused by: java.nio.BufferUnderflowException
	at java.nio.Buffer.nextGetIndex(Buffer.java:500)
	at java.nio.HeapByteBuffer.get(HeapByteBuffer.java:135)
	at com.esotericsoftware.kryo.Serializer.readObject(Serializer.java:54)
	at com.esotericsoftware.kryo.serialize.FieldSerializer.readObjectData(FieldSerializer.java:206)
	at com.esotericsoftware.kryo.serialize.FieldSerializer.readObjectData(FieldSerializer.java:178)
	at com.esotericsoftware.kryo.Kryo.readObjectData(Kryo.java:512)
	... 8 more
```

[This error has been seen before!](https://webarchive.jira.com/browse/HER-1996). 

The BDB-JE documentation notes that all writes to the database should be halted during the backup process. So, one issue is whether the `DbBackup` start/stop enforces that, or whether some H3 threads could be interfering.

A separate issue is whether this is correctly synchronised with the caching mechanism. The clever `ObjectIdentityBdbManualCache` caches a subset of the databases in memory, and syncs on explicit `sync` and when items are evicted from the cache.  Note in particular that if the key is new, a ['mock' entry is dropped into the diskMap](https://github.com/internetarchive/heritrix3/blob/05811705ed996122bea1f4e034c1ed5f7a07240f/commons/src/main/java/org/archive/util/ObjectIdentityBdbManualCache.java#L237-L240).

If the sync/flush to disk is not complete and the whole system is not locked during checkpointing, this could cause problems.  

For the frontier, [this supplies the initial entry](https://github.com/internetarchive/heritrix3/blob/05811705ed996122bea1f4e034c1ed5f7a07240f/engine/src/main/java/org/archive/crawler/frontier/BdbFrontier.java#L151-L157). This sets up the work queue.

However, I think the actual CrawlURIs are stored separately in the same DB?

Yes, the logic in `BdbMultipleWorkQueues` handles things very differently, managing the deferred URIs in a separate section of the same `BdbModule` as the `WorkQueue` stuff.

Really not clear how too-small or empty CrawlURI entries could make it in. Possibly some race condition? Or some entry size limit we've not accounted for.

Anyway, the short-term resolution would appear to be to extend this part of the code to catch the `BufferUnderflowException` and log useful information to help debug the root cause in the future.

https://github.com/internetarchive/heritrix3/blob/aa705bef2eb8fbbb9aeb56995e713a7b6ba0ed00/engine/src/main/java/org/archive/crawler/frontier/BdbMultipleWorkQueues.java#L266-L282

FWIW, the crawl appears to keep going once these entries have been processed. The main problem is that it kills the `ToeThreads`.

**Comment by BitBaron** [comment]
I recently had this issue when trying to stop a checkpointed crawl. Several of the URIs became enqueued and prevented the crawl from terminating.

Upon researching the previous logs I found the URIs had been affected by a change to the [Base32 encoding class](#304). This caused the ToeThread working on these CrawlURIs to terminate early.

```java
Feb 11, 2020 8:30:17 PM org.archive.crawler.framework.ToeThread seriousError
SEVERE: Serious error occurred trying to process 'CrawlURI http://hostspinburneu.cf/ IX https://tagan.adlightning.com/buzzfeed/bl-1ee266c-762cd020.js in org.archive.modules.recrawl.ContentDigestHistoryLoader#2a0b0cd'
[ToeThread #4: http://hostspinburneu.cf/
 CrawlURI http://hostspinburneu.cf/ IX https://tagan.adlightning.com/buzzfeed/bl-1ee266c-762cd020.js    2 attempts
    in processor: org.archive.modules.recrawl.ContentDigestHistoryLoader#2a0b0cd
    ACTIVE for 5s818ms
    step: ABOUT_TO_BEGIN_PROCESSOR for 0ms
Java Thread State: RUNNABLE
Blocked/Waiting On: NONE
    java.lang.Thread.getStackTrace(Thread.java:1559)
    org.archive.crawler.framework.ToeThread.reportThread(ToeThread.java:484)
    org.archive.crawler.framework.ToeThread.reportTo(ToeThread.java:456)
    org.archive.util.DevUtils.extraInfo(DevUtils.java:61)
    org.archive.crawler.framework.ToeThread.seriousError(ToeThread.java:267)
    org.archive.crawler.framework.ToeThread.run(ToeThread.java:177)
]
           timestamp  discovered      queued   downloaded       doc/s(avg)  KB/s(avg)   dl-failures   busy-thread   mem-use-KB  heap-size-KB   congestion   max-depth   avg-depth
2020-02-11T20:30:17Z      151986      115262        43592          0(0.73)      0(79)            20            11       361454       1915904        82.68          91          36
 (in thread 'ToeThread #4: http://hostspinburneu.cf/'; in processor 'org.archive.modules.recrawl.ContentDigestHistoryLoader#2a0b0cd')
java.lang.NoClassDefFoundError: com/google/common/io/BaseEncoding
        at org.archive.util.Base32.encode(Base32.java:33)
        at org.archive.modules.CrawlURI.getContentDigestString(CrawlURI.java:984)
        at org.archive.modules.CrawlURI.getContentDigestSchemeString(CrawlURI.java:968)
        at org.archive.modules.recrawl.AbstractContentDigestHistory.persistKeyFor(AbstractContentDigestHistory.java:53)
        at org.archive.modules.recrawl.TroughContentDigestHistory.load(TroughContentDigestHistory.java:127)
        at org.archive.modules.recrawl.ContentDigestHistoryLoader.innerProcess(ContentDigestHistoryLoader.java:46)
        at org.archive.modules.Processor.innerProcessResult(Processor.java:175)
        at org.archive.modules.Processor.process(Processor.java:142)
        at org.archive.modules.ProcessorChain.process(ProcessorChain.java:131)
        at org.archive.crawler.framework.ToeThread.run(ToeThread.java:152)
```

**Comment by hennekey** [comment]
I tried to reproduce the error from the contrib module with a simple test using the code prior to #312 and was unsuccessful. Any chance you can provide a dependency tree for your project before #312?

**Comment by anjackson** [comment]
I believe this issue was to do with problems in managing checkpoints, and should have been resolved by #281.

**Comment by anjackson** [comment]
I haven't seen this since #281 so I'm calling this fixed. Please re-open if needed.
